Inform type in messsage: 'Illegal offset type'

marcosmarcolin opened the pull request:
https://github.com/php/php-src/issues/10432

@Girgias @nielsdos @devnexen  

I apologize for my ignorance on https://github.com/php/php-src/pull/10450, it was not something I used in my day to day (git rebase).

I appreciate all of your help.

nielsdos commented on ext/opcache/jit/zend_jit_helpers.c:
```diff
@@ -27,9 +27,14 @@ static ZEND_COLD void undef_result_after_exception(void) {
 }
 
-static ZEND_COLD void zend_jit_illegal_offset(void)
+static ZEND_COLD void zend_jit_illegal_array_offset(zval *offset)
```
this offset pointer can become const too: `const zval *offset`.

marcosmarcolin replied:
Thanks for the correction!

nielsdos commented on Zend/zend_execute.c:
```diff
@@ -1466,9 +1466,19 @@ static zend_never_inline ZEND_COLD void ZEND_FASTCALL zend_use_object_as_array(v
+static zend_never_inline ZEND_COLD void ZEND_FASTCALL zend_illegal_unset_offset(const zval *offset)
 {
-	zend_type_error("Illegal offset type");
+	zend_type_error("Cannot access offset of type %s on unset", zend_get_type_by_const(Z_TYPE_P(offset)));
```
typo: `on unset` -> `in unset`

marcosmarcolin replied:
Thanks for the correction!

nielsdos commented on ext/spl/spl_array.c:
```diff
@@ -89,6 +89,21 @@ static inline HashTable **spl_array_get_hash_table_ptr(spl_array_object* intern)
+
+static void spl_array_illegal_unset_offset(const zval *offset)
+{
+	zend_type_error("Cannot access offset of type %s on unset", zend_get_type_by_const(Z_TYPE_P(offset)));
```
Same typo as before: `on unset` -> `in unset`

marcosmarcolin replied:
Thanks for the correction!

nielsdos reviewed (commented):
I had a quick scroll throug the changed C code and it looks generally good, thanks.
Just some minor comments.

mvorisek commented on Zend/tests/036.phpt:
```diff
@@ -11,4 +11,4 @@ try {
 ?>
 --EXPECT--
-Illegal offset type
+Cannot access offset of type object on array
```
I propose `Illegal array offset of type object` to stress the type (as array offset) is not supported.

devnexen replied:
I prefer too.

Girgias replied:
Can we have this as a *follow* up PR as this is the wording used for string offsets:
```
zend_type_error("Cannot access offset of type %s on string", zend_zval_value_name(offset));
```

The whole exercise of this PR is to align array offset and string offset error messages.

mvorisek replied:
"Illegal offset type" was always present and this PR this to describe the error newly with "in unset" (and friends), so I belive the "Cannot access offset" should not be introduced in the first place.

Girgias reviewed (approved):
Other than the typo looks good.

I'll do a follow up PR to extract the offset functions into a common zend_API one as it could also fold the string offset access error into it.

Girgias commented:
After addressing the comments please do not forget to regenerate the vm by running ``Zend/zend_vm_gen.php``

nielsdos commented:
Lgtm, thanks!

marcosmarcolin commented:
> After addressing the comments please do not forget to regenerate the vm by running Zend/zend_vm_gen.php

Done, I think that's it. I'm for you.

Thanks.

marcosmarcolin commented:
About the failure in the CI, I noticed the same behavior in https://github.com/php/php-src/pull/10483.

Appears to be cached.

marcosmarcolin commented:
Well, now it's over, it was the same cache. The `AppVeyor` failure I believe is related to this change.

Girgias commented:
> Well, now it's over, it was the same cache. The `AppVeyor` failure I believe is related to this change.

The failed file is ``file_get_contents_variation7-win32.phpt`` and it doesn't access any offsets.

Girgias commented:
Thank you!

marcosmarcolin commented:
> The failed file is file_get_contents_variation7-win32.phpt and it doesn't access any offsets.

Okay, that was a typo, it was supposed to say it had nothing to do with this PR changes.

Thanks for all your help, regards.
